flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStore.java

```diff
@@ -19,92 +19,224 @@
+
+	/** Lock for synchronisation on the storage map and the discarded status. */
+	private final Object lock;
+
+	/** Status flag if this store was already discarded. */
+	@GuardedBy("lock")
+	private boolean discarded;
+
+	/** Maps checkpoint ids to local TaskStateSnapshots. */
+	@GuardedBy("lock")
+	private final SortedMap<Long, TaskStateSnapshot> storedTaskStateByCheckpointID;
+
 	public TaskLocalStateStore(
-		JobID jobID,
-		JobVertexID jobVertexID,
-		int subtaskIndex,
-		File[] localStateRootDirectories) {
+		@Nonnull JobID jobID,
+		@Nonnull JobVertexID jobVertexID,
+		@Nonnegative int subtaskIndex,
+		@Nonnull File[] localStateRootDirectories,
+		@Nonnull Executor discardExecutor) {
 
 		this.jobID = jobID;
 		this.jobVertexID = jobVertexID;
 		this.subtaskIndex = subtaskIndex;
-		this.storedTaskStateByCheckpointID = new HashMap<>();
 		this.localStateRootDirectories = localStateRootDirectories;
+		this.discardExecutor = discardExecutor;
+		this.lock = new Object();
+		this.storedTaskStateByCheckpointID = new TreeMap<>();
+		this.discarded = false;
 	}
 
+	@Nonnull
 	protected String createSubtaskPath() {
 		return jobID + File.separator + jobVertexID + File.separator + subtaskIndex;
 	}
 
+	/**
+	 * Stores the local state for the given checkpoint id.
+	 *
+	 * @param checkpointId id for the checkpoint that created the local state that will be stored.
+	 * @param localState the local state to store.
+	 */
 	public void storeLocalState(
-		@Nonnull CheckpointMetaData checkpointMetaData,
+		@Nonnegative long checkpointId,
 		@Nullable TaskStateSnapshot localState) {
 
-		TaskStateSnapshot previous =
-			storedTaskStateByCheckpointID.put(checkpointMetaData.getCheckpointId(), localState);
+		if (localState == null) {
+			localState = NULL_DUMMY;
+		}
 
-		if (previous != null) {
-			throw new IllegalStateException("Found previously registered local state for checkpoint with id " +
-				checkpointMetaData.getCheckpointId() + "! This indicated a problem.");
+		LOG.info("Storing local state for checkpoint {}.", checkpointId);
+		LOG.debug("Local state for checkpoint {} is {}.", checkpointId, localState);
+
+		synchronized (lock) {
+			if (discarded) {
+				// we ignore late stores and simply discard the state.
+				discardStateObject(localState, checkpointId);
+			} else {
+				TaskStateSnapshot previous =
+					storedTaskStateByCheckpointID.put(checkpointId, localState);
+
+				if (previous != null) {
+					// this should never happen.
+					discardStateObject(previous, checkpointId);
+					throw new IllegalStateException("Found previously registered local state for checkpoint with id " +
+						checkpointId + "! This indicated a problem.");
+				}
+
+				// prune history.
+				while (storedTaskStateByCheckpointID.size() > MAX_RETAINED_SNAPSHOTS) {
+					Long removeCheckpointID = storedTaskStateByCheckpointID.firstKey();
+					TaskStateSnapshot snapshot =
+						storedTaskStateByCheckpointID.remove(storedTaskStateByCheckpointID.firstKey());
```

Comment:
I think we can directly pass in `removeCheckpointID`.

Reply:
Deleted in later commit.